src/diffusers/models/modeling_utils.py

```diff
@@ -803,7 +803,7 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                             model._temp_convert_self_to_deprecated_attention_blocks()
                             accelerate.load_checkpoint_and_dispatch(
                                 model,
-                                model_file if not is_sharded else sharded_ckpt_cached_folder,
+                                model_file if not is_sharded else index_file,
```

Comment:
Can we also add a test for variant loading as well? 

Ultimately we should be able to load https://huggingface.co/fal/AuraFlow/tree/main/transformer with `variant="fp16"`. 

Would be nice if you could test and report?

Reply:
Sounds good ! Could you just give me the snippet you used to generate these models ? https://huggingface.co/hf-internal-testing/unet2d-sharded-dummy/tree/main . I will create one repo that contain both the fp32 version and the fp16 version ? 

Also regarding the auraflow model, by running the following: 
```python
from diffusers import AuraFlowTransformer2DModel
model = AuraFlowTransformer2DModel.from_pretrained("fal/AuraFlow", variant="fp16", subfolder="transformer")
```
I am able to download + load the model with the correct variant: 
```
transformer/config.json: 100%|█████████████████████████████| 511/511 [00:00<00:00, 138kB/s]
(…)ytorch_model.safetensors.fp16.index.json: 100%|█████| 37.7k/37.7k [00:00<00:00, 461kB/s]
(…)ch_model-00002-of-00002.fp16.safetensors: 100%|████| 3.77G/3.77G [02:05<00:00, 30.1MB/s]
(…)ch_model-00001-of-00002.fp16.safetensors: 100%|████| 9.91G/9.91G [20:23<00:00, 8.10MB/s]
Fetching 2 files: 100%|█████████████████████████████████████| 2/2 [20:23<00:00, 611.97s/it]
['/home/marc/.cache/huggingface/hub/models--fal--AuraFlow/snapshots/f16d500f22e8689f10581936c6a453d7c6fe646a/transformer/diffusion_pytorch_model-00001-of-00002.fp16.safetensors', '/home/marc/.cache/huggingface/hub/models--fal--AuraFlow/snapshots/f16d500f22e8689f10581936c6a453d7c6fe646a/transformer/diffusion_pytorch_model-00002-of-00002.fp16.safetensors']
```

Reply:
@sayakpaul just curious, for aura, how do they currently load these fp16 checkpoints?

Reply:
So, that is the thing. If you try specifying `variant="fp16"`, it will load the non-fp16 variants, incorrectly. That is what we are fixing here. 

Note that this doesn't happen for non-sharded checkpoints.

Reply:
Perfect! Thanks, @SunMarc! 

> Could you just give me the snippet you used to save this model ?

So, I just add a `model.save_pretrained()` here:
https://github.com/huggingface/diffusers/blob/52f1378e64367c655461d4b6d5b2e19585b92791/tests/models/test_modeling_common.py#L890

And then run https://github.com/huggingface/diffusers/blob/main/tests/models/unets/test_models_unet_2d_condition.py with `-k "test_sharded_checkpoints"` in `pytest`. 

In this case, be sure to provide `variant="fp16"` in `save_pretrained()`. I am sure you know that already but just wanted to mention.

Reply:
Thanks for the advice ! I've uploaded the model [here](https://huggingface.co/hf-internal-testing/unet2d-sharded-with-variant-dummy/tree/main) and added a test !

Reply:
I think you forgot to push the test?

Reply:
Oh yes, I pushed on the wrong remote. Should be fixed now